### PR TITLE
[REVIEW] Add `cudf::encode`

### PR DESCRIPTION
**Pull request by shwina**
Essentially moves functionality out of `encode.cu` into a public API and adds tests for it. See #5498.

### CODE REVIEW

**Comment by GPUtester** [comment]
Please update the changelog in order to start CI tests.

View the gpuCI docs [here](https://docs.rapids.ai/gpuci#open-pr-or-update-pr).

File: cpp/src/transform/codify.cpp [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+      mr);
+    keys_column->set_null_mask(rmm::device_buffer{0, stream, mr}, 0);  // remove the null-mask
+  }
+
+  // this returns a column with no null entries
+  // - it appears to ignore the null entries in the input and tries to place the value regardless
+  auto indices_column = cudf::detail::lower_bound(table_view{{keys_column->view()}},
+                                                  table_view{{input_column}},
+                                                  std::vector<order>{order::ASCENDING},
+                                                  std::vector<null_order>{null_order::AFTER},
```
Comment (shwina) [inline cpp/src/transform/codify.cpp:42]: We should make null placement configurable. One question to consider:

Let's say we want to codify {'a', 'b', null, 'b', 'a'}. There are three possible answers:

1. `{0, 1, 2, 1, 0}` (nulls come last)
2. `{1, 2, 0, 1, 2}` (nulls come first)
3. `{0, 1, null, 1, 0}` (nulls are not counted at all)

Is (3) interesting enough to support in libcudf? Or are we OK with telling the user to call `replace` as a post-processing step?

File: cpp/include/cudf/transform.hpp [hunk 2]
```diff
@@ -83,5 +83,29 @@ std::pair<std::unique_ptr<rmm::device_buffer>, size_type> nans_to_nulls(
...
+ * where `keys` is the set of distinct values in `input` in sorted order.
+ *
+ * Examples:
+ * input: {'a', 'b', 'b', 'a'}
+ * output: [{'a', 'b'}, {0, 1, 1, 0}]
+ *
+ * input: {1, 3, 1, 2, 9}
+ * output: [{1, 2, 3, 9}, {0, 2, 0, 1, 3}]
+ *
+ * @param input        Column containing values to be encode
```
Comment (harrism) [inline cpp/include/cudf/transform.hpp:101]:
```suggestion
 * @param input        Column containing values to be encoded
```

File: cpp/include/cudf/transform.hpp [hunk 3]
```diff
@@ -83,5 +83,29 @@ std::pair<std::unique_ptr<rmm::device_buffer>, size_type> nans_to_nulls(
...
+ * @param mr           Device memory resource used to allocate the returned bitmask.
```
Comment (harrism) [inline cpp/include/cudf/transform.hpp:102]:
```suggestion
 * @param mr           Device memory resource used to allocate the returned columns.
```

**Comment by shwina** [comment]
Yes - this PR is just exposing functionality already written for dictionary columns. I guess we could just cast the input column to a dictionary column and then grab the indices from there, but it feels like an implementation detail that the keys will always be 0 to N-1.

**Comment by shwina** [comment]
Should have xref'd issue: https://github.com/rapidsai/cudf/issues/5498

File: cpp/src/transform/codify.cpp (hunk 1 above)
Comment (shwina) [reply cpp/src/transform/codify.cpp]: > We should make null placement configurable

Should we?

File: cpp/src/transform/encode.cpp [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+#include <cudf/detail/transform.hpp>
+#include <cudf/table/table_view.hpp>
+
+namespace cudf {
+namespace detail {
+
+std::pair<std::unique_ptr<column>, std::unique_ptr<column>> encode(
+  column_view const& input_column, rmm::mr::device_memory_resource* mr, cudaStream_t stream)
+{
+  // side effects of this function were are now dependent on:
```
Comment (davidwendt) [inline cpp/src/transform/encode.cpp:15]:
```suggestion
  // side effects of this function we are now dependent on:
```

File: cpp/src/transform/encode.cpp [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+  // - resulting column elements are sorted ascending
+  // - nulls are sorted to the beginning
+  auto table_keys = cudf::detail::drop_duplicates(table_view{{input_column}},
+                                                  std::vector<size_type>{0},
+                                                  duplicate_keep_option::KEEP_FIRST,
+                                                  null_equality::EQUAL,
+                                                  mr,
+                                                  stream)
+                      ->release();  // true == nulls are equal
```
Comment (davidwendt) [inline cpp/src/transform/encode.cpp:24]:
```suggestion
                      ->release();
```
This was caused by a reformatting all the code using clang-format some time ago

File: cpp/tests/encode/encode_tests.cpp [hunk 6]
```diff
@@ -0,0 +1,121 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
```
Comment (davidwendt) [inline cpp/tests/encode/encode_tests.cpp:2]:
```suggestion
 * Copyright (c) 2020, NVIDIA CORPORATION.
```

File: cpp/tests/encode/encode_tests.cpp [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include <cudf/transform.hpp>
+#include <tests/utilities/base_fixture.hpp>
+#include <tests/utilities/column_utilities.hpp>
+#include <tests/utilities/column_wrapper.hpp>
+#include <tests/utilities/table_utilities.hpp>
```
Comment (davidwendt) [inline cpp/tests/encode/encode_tests.cpp:20]: Is `tests/utilities/table_utilities.hpp` needed?

File: cpp/tests/CMakeLists.txt [hunk 8]
```diff
@@ -573,7 +573,14 @@ set(DICTIONARY_TEST_SRC
     "${CMAKE_CURRENT_SOURCE_DIR}/dictionary/set_keys_test.cpp"
     "${CMAKE_CURRENT_SOURCE_DIR}/dictionary/slice_test.cpp")
 
-ConfigureTest(DICTIONARY_TEST "${DICTIONARY_TEST_SRC}")
+
+###################################################################################################
+# - codfy tests -----------------------------------------------------------------------------------
```
Comment (davidwendt) [inline cpp/tests/CMakeLists.txt:578]:
```suggestion
# - encode tests -----------------------------------------------------------------------------------
```

File: cpp/tests/encode/encode_tests.cpp (hunk 7 above)
Comment (shwina) [reply cpp/tests/encode/encode_tests.cpp:20]: Removed!

File: cpp/src/transform/codify.cpp (hunk 1 above)
Comment (shwina) [reply cpp/src/transform/codify.cpp]: Could be a post-processing step. Marking resolved.

File: cpp/include/cudf/transform.hpp [hunk 9]
```diff
@@ -83,5 +83,29 @@ std::pair<std::unique_ptr<rmm::device_buffer>, size_type> nans_to_nulls(
...
 
+/**
+ * @brief Encode the values of the given column as integers
+ *
+ * The encoded values are integers in the range [0, n), where `n`
+ * is the number of distinct values in the input column.
+ * The result column is such that keys[result[i]] == input[i],
+ * where `keys` is the set of distinct values in `input` in sorted order.
+ *
+ * Examples:
```
Comment (davidwendt) [inline cpp/include/cudf/transform.hpp:94]: Surround examples with `@code{.pseudo}` and `@endcode` doxygen tags.
https://github.com/rapidsai/cudf/blob/branch-0.15/cpp/docs/DOCUMENTATION.md#inline-examples

File: cpp/include/cudf/detail/transform.hpp [hunk 10]
```diff
@@ -52,5 +52,11 @@ std::pair<std::unique_ptr<rmm::device_buffer>, cudf::size_type> bools_to_mask(
   column_view const& input,
   rmm::mr::device_memory_resource* mr = rmm::mr::get_default_resource(),
   cudaStream_t stream                 = 0);
+
```
Comment (rgsl888prabhu) [inline cpp/include/cudf/detail/transform.hpp:55]: doc?

File: cpp/include/cudf/transform.hpp [hunk 11]
```diff
@@ -83,5 +83,29 @@ std::pair<std::unique_ptr<rmm::device_buffer>, size_type> nans_to_nulls(
 std::pair<std::unique_ptr<rmm::device_buffer>, cudf::size_type> bools_to_mask(
   column_view const& input, rmm::mr::device_memory_resource* mr = rmm::mr::get_default_resource());
 
+/**
+ * @brief Encode the values of the given column as integers
+ *
+ * The encoded values are integers in the range [0, n), where `n`
```
Comment (jrhemstad) [inline cpp/include/cudf/transform.hpp:89]: Describe the behavior for nulls in `input`.

File: cpp/include/cudf/transform.hpp (hunk 9 above)
Comment (shwina) [reply cpp/include/cudf/transform.hpp:94]: Nice! Done

File: cpp/include/cudf/transform.hpp (hunk 11 above)
Comment (shwina) [reply cpp/include/cudf/transform.hpp:89]: Just added a description - let me know if it's clear enough

File: cpp/include/cudf/transform.hpp [hunk 12]
```diff
@@ -83,5 +83,29 @@ std::pair<std::unique_ptr<rmm::device_buffer>, size_type> nans_to_nulls(
...
+ * is the number of distinct values in the input column.
+ * The result column is such that keys[result[i]] == input[i],
+ * where `keys` is the set of distinct values in `input` in sorted order.
```
[the remark this answers is not in the document]
Comment (shwina) [reply cpp/include/cudf/transform.hpp]: fixed

File: cpp/include/cudf/transform.hpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (shwina) [reply cpp/include/cudf/transform.hpp]: fixed

File: cpp/include/cudf/detail/transform.hpp (hunk 10 above)
Comment (shwina) [reply cpp/include/cudf/detail/transform.hpp:55]: Nice catch -- fixed

File: cpp/include/cudf/transform.hpp (hunk 11 above)
Comment (shwina) [reply cpp/include/cudf/transform.hpp:89]: Ah but it's not really accurate.

Comment (shwina) [reply cpp/include/cudf/transform.hpp:89]: Fixed

**Pull request merged**